src/main/java/org/prebid/server/bidder/flipp/FlippBidder.java

```diff
@@ -0,0 +1,270 @@
+        return URLEncodedUtils.parse(url, StandardCharsets.UTF_8);
+    }
+
+    private static String resolveIp(BidRequest bidRequest, ExtImpFlipp extImpFlipp) {
+        if (StringUtils.isNoneEmpty(extImpFlipp.getIp())) {
+            return extImpFlipp.getIp();
+        }
+
+        return Optional.ofNullable(bidRequest.getDevice())
+                .map(Device::getIp)
+                .filter(StringUtils::isNoneEmpty)
+                .orElseThrow(() -> new PreBidException("No IP set in Flipp bidder params or request device"));
+    }
+
+    private static String resolveKey(BidRequest bidRequest, ExtImpFlipp extImpFlipp) {
+        final String userId = ObjectUtil.getIfNotNull(bidRequest.getUser(), User::getId);
+        if (StringUtils.isNoneEmpty(userId)) {
+            return userId;
+        }
+
+        return Optional.ofNullable(extImpFlipp.getUserKey())
+                .filter(StringUtils::isNotEmpty)
+                .orElseGet(() -> UUID.randomUUID().toString());
+    }
+
+    private HttpRequest<CampaignRequestBody> createRequest(BidRequest bidRequest,
+                                                           CampaignRequestBody campaignRequestBody) {
+        return HttpRequest.<CampaignRequestBody>builder()
+                .method(HttpMethod.POST)
+                .uri(endpointUrl)
+                .headers(updateHeaders(bidRequest.getDevice()))
+                .body(mapper.encodeToBytes(campaignRequestBody))
+                .payload(campaignRequestBody)
+                .build();
+    }
+
+    private static MultiMap updateHeaders(Device device) {
+        return Optional.of(device)
+                .map(Device::getUa)
+                .map(ua -> HttpUtil.headers().add(HttpUtil.USER_AGENT_HEADER, ua))
+                .orElseGet(HttpUtil::headers);
+    }
+
+    @Override
+    public final Result<List<BidderBid>> makeBids(BidderCall<CampaignRequestBody> httpCall, BidRequest bidRequest) {
+        try {
+            final CampaignResponseBody campaignResponseBody =
+                    mapper.decodeValue(httpCall.getResponse().getBody(), CampaignResponseBody.class);
+            return Result.withValues(extractInline(campaignResponseBody, bidRequest));
+        } catch (DecodeException e) {
+            return Result.withError(BidderError.badServerResponse(e.getMessage()));
+        }
+    }
+
+    private static List<BidderBid> extractInline(CampaignResponseBody campaignResponseBody, BidRequest bidRequest) {
+        if (campaignResponseBody == null
+                || Objects.isNull(campaignResponseBody.getDecisions())
+                || CollectionUtils.isEmpty(campaignResponseBody.getDecisions().getInline())) {
+            return Collections.emptyList();
+        }
+
+        return bidsFromResponse(campaignResponseBody, bidRequest);
+    }
+
+    private static List<BidderBid> bidsFromResponse(CampaignResponseBody campaignResponseBody, BidRequest bidRequest) {
```

Comment:
looks line here you don't need full `CampaignResponseBody` just inlines will be enough